src/replay/consensus/core.zig

```diff
@@ -878,6 +974,407 @@ fn pushVote(
+    defer seen_sockets.deinit(allocator);
+
+    var leader_sockets: std.ArrayListUnmanaged(SocketAddr) = .empty;
+    defer leader_sockets.deinit(allocator);
+
+    var leader_iter = seen_leaders.keyIterator();
+    while (leader_iter.next()) |leader_pubkey_ptr| {
+        const leader_pubkey = leader_pubkey_ptr.*;
+        const contact_info = gossip_table.getThreadSafeContactInfo(leader_pubkey) orelse continue;
+
+        const socket_addr = contact_info.tpu_addr orelse continue;
+
+        const gop = try seen_sockets.getOrPut(allocator, socket_addr);
+        if (!gop.found_existing) {
+            try leader_sockets.append(allocator, socket_addr);
+        }
+    }
+
+    return leader_sockets.toOwnedSlice(allocator);
+}
+
+fn sendVoteTransaction(
+    logger: Logger,
+    allocator: Allocator,
+    vote_tx: Transaction,
+    tpu_address: SocketAddr,
+) !void {
+    var serialized: std.ArrayListUnmanaged(u8) = .empty;
+    defer serialized.deinit(allocator);
+
+    try sig.bincode.write(serialized.writer(allocator), vote_tx, .{});
+
+    const socket = switch (tpu_address) {
+        .V4 => network.Socket.create(.ipv4, .udp) catch |err| {
+            logger.err().logf("Failed to create socket for vote send: {}", .{err});
+            return err;
+        },
+        .V6 => network.Socket.create(.ipv6, .udp) catch |err| {
+            logger.err().logf("Failed to create socket for vote send: {}", .{err});
+            return err;
+        },
+    };
+    defer socket.close();
+
+    const endpoint = tpu_address.toEndpoint();
+    _ = socket.sendTo(endpoint, serialized.items) catch |err| {
+        logger.err().logf("Failed to send vote transaction: {}", .{err});
+        return err;
+    };
+}
+
+fn sendVoteToLeaders(
+    logger: Logger,
+    allocator: Allocator,
+    vote_slot: Slot,
+    vote_tx: Transaction,
+    leader_schedule: *sig.core.leader_schedule.LeaderScheduleCache,
+    gossip_table_rw: *sig.sync.RwMux(sig.gossip.GossipTable),
+) !void {
+    const UPCOMING_LEADER_FANOUT_SLOTS: u64 =
+        FORWARD_TRANSACTIONS_TO_LEADER_AT_SLOT_OFFSET + 1;
+
+    const upcoming_leader_sockets = try upcomingLeaderTpuVoteSockets(
+        allocator,
+        vote_slot,
+        leader_schedule,
+        gossip_table_rw,
+        UPCOMING_LEADER_FANOUT_SLOTS,
+    );
+    defer allocator.free(upcoming_leader_sockets);
+
+    for (upcoming_leader_sockets) |tpu_vote_socket| {
+        sendVoteTransaction(logger, allocator, vote_tx, tpu_vote_socket) catch |err| {
+            logger.err().logf("Failed to send vote to leader: {}", .{err});
+        };
+    }
+}
+
+fn sendVoteToGossip(
+    vote_op: VoteOp,
+    gossip_table_rw: *sig.sync.RwMux(sig.gossip.GossipTable),
+    my_keypair: sig.identity.KeyPair,
+) !void {
+    const gossip_table, var gossip_table_lock = gossip_table_rw.writeWithLock();
+    defer gossip_table_lock.unlock();
+
+    const now = sig.time.getWallclockMs();
```

Comment:
Would prefer to make this a parameter that can be sent down from `sendVote`, so tests can supply this as a deterministic value.